events/app.py

```diff
@@ -0,0 +1,154 @@
+import os
+import logging
+import datetime
+import pymongo
+
+from collections import namedtuple
+from flask import Flask, request, Response
+
+app = Flask(__name__)
+
+
+def connect_to_mongodb():
+    """Connects to MongoDB Atlas database.
+
+    Returns events collection if connection is successful, and None otherwise.
+    """
+    mongodb_uri = os.environ.get("MONGODB_URI")
+    if mongodb_uri is None:
+        print("Alert: not able to find MONGODB_URI environmental variable, "
+              "no connection to MongoDB instance")
+        return None  # not able to find db config var
+    return pymongo.MongoClient(mongodb_uri).eventsDB.all_events
+
+
+EVENTS_COLL = connect_to_mongodb()  # None if can't connect
+
+
+@app.route('/v1/add', methods=['POST'])
+def add_event():
+    # TODO(cmei4444): Get event information from request data
+    info = request.form['info']
+    # TODO(cmei4444): Verify that user has event editing access
+    current_time = datetime.datetime.now()
+    build_event_info(info, current_time)
+    try:
+        event = Event(info)
+    except ValueError:      # missing or extra event attributes
+        return Response(
+            status=400,
+            response="Event info was entered incorrectly.",
+        )
+    if EVENTS_COLL is not None:
+        event.add_to_db(EVENTS_COLL)
+    else:
+        return Response(
+            status=500,
+            response="Database was undefined.",
+        )
+    return Response(
+        status=201,
+    )
+
+
+def build_event_info(info, time):
+    """Adds created_at time to event info dict."""
+    info['created_at'] = time
+    return info
+
+
+@app.route('/v1/edit/<event_id>', methods=['PUT'])
+def edit_event(event_id):
+    """Edit the event with the given id."""
+    pass
+
+
+@app.route('/v1/', methods=['GET'])
+def get_all_events():
+    """Return a list of all events currently in the DB."""
+    pass
+
+
+@app.route('/v1/search', methods=['GET'])
+def search_event():
+    """Search for the event with the given name in the DB."""
+    pass
+
+
+@app.route('/v1/<event_id>', methods=['PUT'])
+def get_one_event(event_id):
+    """Retrieve one event by event_id."""
+    pass
+
+
+def equal_times(time_1, time_2):
+    """Determines if two times are equal.
+
+    Times can be represented as strings (mainly for testing) or as
+    datetime.datetime objects. Datetime objects are rounded to the nearest
+    millisecond before comparison.
+
+    Used for comparing Event objects, since MongoDB truncates times to the
+    nearest millisecond.
+    """
+    if isinstance(time_1, type(time_2)):
+        if isinstance(time_1, datetime.datetime):
+            time_1 = time_1.replace(
+                microsecond=(time_1.microsecond // 1000) * 1000)
+            time_2 = time_2.replace(
+                microsecond=(time_2.microsecond // 1000) * 1000)
+        return time_1 == time_2
```

Comment:
Wow, that's so much cleaner! The more you know.

Reply:
It looks like `//` is actually integer division on timedeltas (from the link you included) so this seems to not work.

Reply:
That's unfortunate. Why again do we have to round to the nearest millisecond before comparison? Are you expecting the same event to be recorded in multiple places with times nanoseconds apart but that should be considered the same?

Reply:
It appears that mongodb stores millisecond-level precision, but the default for python is microsecond-level. This means that the microseconds number will be "wrong" (not-equal) after a round-trip through the database.

Unfortunately, it looks like @cmei4444 is correct, and the `//` operator only works on timedeltas (and returns an integer when it does so).

This looks about as clean as any other option; if we need to use this more, we should probably create a `monogodb_time` function to strip the milliseconds.

Reply:
I expect we will also need this for the posts service, so to follow DRY we should probably make some sort of common library for the whole project then.

Reply:
A different route we could take would be to store all the times as strings in some agreed-upon format (I suggested one in an earlier comment), which is pretty easy to convert back and forth to datetimes

Reply:
> A different route we could take would be to store all the times as strings in some agreed-upon format (I suggested one in an earlier comment), which is pretty easy to convert back and forth to datetimes

I like that idea a lot better than all this datetime madness, especially because if we generate the string like you suggested earlier we can just not include the microsecond part and easily truncate it. I imaging the 2 things we really want these date times recorded for is to order events/posts by time and to give a client-side label on events/posts for when they were posted, both of which are easily accomplished by strings.

Reply:
Zeroing the microseconds value would also work here; if you decide on a string format, you probably want to:

1. Define the time zone that the times are in (you probably want UTC)
2. Ensure that times are consistently ascending (i.e. as the ISO timestamp format ensures)